source/slang/glsl.meta.slang

```diff
@@ -2825,21 +2825,1988 @@ public uint rayQueryGetIntersectionTypeEXT(rayQueryEXT q, bool committed)
+        if (__type_equals<T, uint64_t>()) __requireGLSLExtension("GL_EXT_shader_subgroup_extended_types_int64");
+        return;
+    case spirv:
+    case cpp:
+    case cuda:
+    case cpp:
+        return;
+    }
+    return;
+} 
 
-//
-// Subgroup
-//
+// GL_KHR_shader_subgroup_basic Built-in Variables 
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_NumSubgroups;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupID;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupSize;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupInvocationID;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupEqMask;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupGeMask;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupGtMask;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupLeMask;
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+public uint gl_SubgroupLtMask;
+
+// GL_KHR_shader_subgroup_basic
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+spirv_caps_shader_subgroup(1.3)
+[ForceInline] public void subgroupBarrier()
+{
+    __target_switch
+    {
+    case cuda:
+        __intrinsic_asm "__syncwarp()";
+    case hlsl:
+        __intrinsic_asm "AllMemoryBarrierWithGroupSync()";
+    case glsl:
+        __intrinsic_asm "subgroupBarrier()";
+    case spirv:
+        spirv_asm {
+            OpCapability Shader;
+            OpControlBarrier Subgroup Subgroup AcquireRelease|SubgroupMemory|ImageMemory|UniformMemory
+        };
+    case cpp:
+        // TODO: cpp
+        return;
+    }
+}
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+spirv_caps_shader_subgroup(1.3)
+[ForceInline] public void subgroupMemoryBarrier()
+{
+    __target_switch
+    {
+    case cuda:
+        __intrinsic_asm "__threadfence_block()";
+    case hlsl:
+        __intrinsic_asm "AllMemoryBarrier()";
+    case glsl:
+        __intrinsic_asm "subgroupMemoryBarrier()";
+    case spirv:
+        spirv_asm {
+            OpCapability Shader;
+            OpMemoryBarrier Subgroup AcquireRelease|SubgroupMemory|ImageMemory|UniformMemory
+        };
+    case cpp:
+        // TODO: cpp
+        return;
+    }
+}
+
+glsl_caps_shader_subgroup(GL_KHR_shader_subgroup_basic)
+spirv_caps_shader_subgroup(1.3)
+[ForceInline] public void subgroupMemoryBarrierBuffer()
+{
+    __target_switch
+    {
+    case cuda:
+        __intrinsic_asm "__threadfence_block()";
+    case hlsl:
+        __intrinsic_asm "DeviceMemoryBarrier()";
+    case glsl:
+        __intrinsic_asm "subgroupMemoryBarrierBuffer()";
+    case spirv:
+        spirv_asm {
+            OpCapability Shader;
+            OpMemoryBarrier Subgroup AcquireRelease|UniformMemory
+        };
+    case cpp:
+        // TODO: cpp
+        return;
+    }
+}
```

Comment:
Can this be like following?
```
[ForceInline] public void subgroupMemoryBarrierBuffer()
{
    return DeviceMemoryBarrier();
}
```

Reply:
subgroupMemoryBarrierBuffer targets syncing within the subgroup, the defined hlsl.meta.slang function targets a larger granularity of barrier. HLSL lacks an equivalent to GLSL & SPIR-V.

Reply:
Probably it will be good idea to leave a comment with the information.
Because any programmer who will look at the function will have the same question as I had.

Reply:
(elaboration added in code comment)